teal/packer.tl

```diff
@@ -45,36 +45,52 @@ local function load_plugin_configs(plugins: {string:Plugin})
   end
 end
 
-local record StartupSpec
-  {UserSpec}
+local did_setup = false
 
-  config: Config
-end
-
--- Convenience function for simple setup
--- spec can be a table with a table of plugin specifications as its first
--- element, config overrides as another element.
-function M.startup(spec: StartupSpec)
+local function setup(user_config: Config)
   local log = require 'packer.log'
-
-  log.debug('start')
-
-  assert(type(spec) == 'table')
-  assert(type(spec[1]) == 'table')
-
-  config(spec.config)
+  log.debug('setup')
+  config(user_config)
 
   for _, dir in ipairs{config.opt_dir, config.start_dir} do
     if vim.fn.isdirectory(dir) == 0 then
       vim.fn.mkdir(dir, 'p')
     end
   end
 
+  did_setup = true
+end
+
+function M.add(spec: UserSpec)
+  if not did_setup then
+    setup()
+  end
+
   local plugin = require 'packer.plugin'
+  load_plugin_configs(plugin.process_spec(spec))
```

Comment:
The intent is `add` is only called once for 90% of users. The other 10% will use it to conditionally specify plugins.

I really don't think there's any need for this. I honestly believe it won't provide any increased performance in the general case. If you think it will then we can revisit this later when we start benchmarking stuff. For now (at least for this PR), I'd prefer we try and design a simple as possible API without potential performance concerns. Most perf gains will be from io and rtp optimizations.